bump(libllvm/flang): 20.1.8 to 21.1.1

finagolfin opened the pull request:
Check if it builds, as I just corrected the patches locally.

robertkirkman commented:
> `error: use of undeclared identifier 'TC'`

sorry if you already are working on it and know what to do, but in case you need help, when I look at the code related to this error, I could suggest changing the block

```c++
if (TC.getTriple().isAndroid()) {
      CmdArgs.push_back("-l:libandroid-complex-math.a");
}
```
into 

```c++
if (getTriple().isAndroid()) {
      CmdArgs.push_back("-l:libandroid-complex-math.a");
}
```

because `getTriple()` is called without an explicit object inside of the source file `ToolChain.cpp` where this is moved to, because in C++ object oriented programming, the original `TC` variable referred to an object of class `ToolChain`, but now that this block is inside a C++ Method which is part of the class `ToolChain`, the use of the method `getTriple()` will automatically run in the context of the current `ToolChain` object, no matter which object it is.

sorry if I overexplained this too much, maybe it is not a difficult problem for you to solve on your own.

finagolfin commented:
Fixed the issue and updated: thanks for looking into it, but I had simply moved that patch so it stayed in the method it was in, without checking if it still needed that qualifier.

robertkirkman commented:
I will try to fix `flang` soon unless licy183 already has an immediate solution, and I will also test my `mesa` PR with this (since `mesa` is another package that requires recompiling when `libllvm` is bumped)

robertkirkman commented:
I've found a workaround that seems correct - the required changes are not yet present in any stable release of `spirv-headers`, so to proceed without waiting, one way that works for me is to cherry-pick this PR and apply it to the `spirv-headers` package version 1.4.321.0, revision-bump that package, and then bump `spirv-llvm-translator`, and then recompile `mesa`.

- https://github.com/KhronosGroup/SPIRV-Headers/pull/532

Everything seems to work fine for me locally after doing that, so I can add my local changes into the PR for bumping `mesa` after this PR has been merged.

truboxl commented:
I can confirm the current rust package can be rebuild to work with LLVM 21 without extra patches.

licy183 commented:
No. It doesn't work properly because `flang-rt` is missing. I'll have a look later.

robertkirkman commented:
I started also trying to help to fix `flang` and unfortunately, it might be very difficult and take a very long time, but I will try to help.

I tried this so far:


```diff
@@ -78,6 +76,21 @@ termux_step_post_configure() {
 }
 
 termux_step_post_make_install() {
+	termux_setup_flang
+	# attempt to build libflang_rt.runtime.a
+	mkdir -p "$TERMUX_PKG_BUILDDIR/flang-rt"
+	cmake -S "$TERMUX_PKG_SRCDIR/runtimes" \
+		-B "$TERMUX_PKG_BUILDDIR/flang-rt" \
+		-GNinja \
+		-DLLVM_DIR=$TERMUX_PREFIX/lib/cmake/llvm \
+		-DCLANG_DIR=$TERMUX_PREFIX/lib/cmake/clang \
+		-DDLLVM_BINARY_DIR="$(dirname $(dirname $(command -v "$FC")))" \
+		-DCMAKE_Fortran_COMPILER="$(command -v "$FC")" \
+		-DCMAKE_Fortran_COMPILER_WORKS=yes \
+		-DLLVM_ENABLE_RUNTIMES=flang-rt
+	
+	cmake --build "$TERMUX_PKG_BUILDDIR/flang-rt" --parallel $TERMUX_PKG_MAKE_PROCESSES
+
 	# Copy module source files
 	mkdir -p $TERMUX_PREFIX/opt/flang/{include,module}
 	cp -f $TERMUX_PKG_SRCDIR/flang/module/* $TERMUX_PREFIX/opt/flang/module/
```

and unfortunately, this happens for me currently:

```
[58/121] Building Fortran object flang-rt/lib/runtime/CMakeFiles/f...er/.termux-build/flang/src/flang/module/iso_fortran_env_impl.f90.o
FAILED: flang-rt/lib/runtime/CMakeFiles/flang_rt.runtime.static.dir/home/builder/.termux-build/flang/src/flang/module/iso_fortran_env_impl.f90.o flang-rt/lib/runtime/iso_fortran_env_impl.mod 
/home/builder/.termux-build/_cache/android-r28c-api-24-v1-flang-v0/bin/aarch64-linux-android-flang-new -I/home/builder/.termux-build/flang/src/flang/module -I/home/builder/.termux-build/flang/src/flang-rt/include -I/home/builder/.termux-build/flang/src/flang-rt/../flang/include -I/home/builder/.termux-build/flang/build/flang-rt/flang-rt -module-dirflang-rt/lib/runtime -fno-lto -U_GLIBCXX_ASSERTIONS -U_LIBCPP_ENABLE_ASSERTIONS -ffixed-line-length-72 -c flang-rt/lib/runtime/CMakeFiles/flang_rt.runtime.static.dir/home/builder/.termux-build/flang/src/flang/module/iso_fortran_env_impl.f90-pp.f90 -o flang-rt/lib/runtime/CMakeFiles/flang_rt.runtime.static.dir/home/builder/.termux-build/flang/src/flang/module/iso_fortran_env_impl.f90.o
error: Semantic errors in flang-rt/lib/runtime/CMakeFiles/flang_rt.runtime.static.dir/home/builder/.termux-build/flang/src/flang/module/iso_fortran_env_impl.f90-pp.f90
/home/builder/.termux-build/flang/src/flang/module/iso_fortran_env_impl.f90:48:53: error: No explicit type declared for 'selected_unsigned_kind'
        integer, parameter, public :: selectedUInt8 = selected_unsigned_ki&
                                                      ^
/home/builder/.termux-build/flang/src/flang/module/iso_fortran_env_impl.f90:66:42: error: No explicit type declared for 'uint'
       &2, -1, selectedUInt8 >= 0), digits(uint(0,kind=safeUInt8)) == 8), &
                                           ^^^^
ninja: build stopped: subcommand failed.
```

I think that a large part of the problem (at least with my attempt) may be that an **`aarch64-linux-android-flang-new`** of **version 21** may be required, and **`aarch64-linux-android-flang-new`** of **version 19** may unfortunately be insufficient to compile the newly-invented `libflang_rt.runtime.so` and/or `libflang_rt.runtime.a`, which **do not seem to have existed prior to LLVM 21**.

```
tacokoneko@CORSAIR ~/code/termux/termux-packages-icewm $ scripts/run-docker.sh 
Running container 'icewm-termux-package-builder' from image 'ghcr.io/termux/package-builder'...
builder@2b24ac818d44:~/termux-packages$ /home/builder/.termux-build/_cache/android-r28c-api-24-v1-flang-v0/bin/aarch64-linux-android-flang-new --version
Android (dev, -pgo, -bolt, -lto, -mlgo, based on r530567e) flang-new version 19.0.1 (https://android.googlesource.com/toolchain/llvm-project 97a699bf4812a18fb657c2779f5296a4ab2694d2)
Target: aarch64-unknown-linux-android24
Thread model: posix
InstalledDir: /home/builder/.termux-build/_cache/android-r28c-api-24-v1-flang-v0/bin
builder@2b24ac818d44:~/termux-packages$ 
```

I have been trying to follow the directions given by this documentation, which has a worrying passage

https://github.com/llvm/llvm-project/blob/main/flang-rt/README.md

> `CMAKE_Fortran_COMPILER` **must be `flang` built from the same Git commit as Flang-RT to ensure they are using the same ABI.** The C and C++ compiler can be any compiler supporting the same ABI.

That is what makes me think that maybe it will be necessary to somehow implement **a `termux_setup_flang_21` which can provide an `aarch64-linux-android-flang-new` for Ubuntu GNU/Linux to compile Fortran targeting Android which reports version 21 instead of version 19**, and I am worried that may be very difficult to create. However it might be possible and I will try.

finagolfin commented:
Bumped this to the latest patch release, test it out and let us know if you see any problems, including with flang.

robertkirkman commented:
These packages are working for me in actual testing after recompiling using LLVM 21.1.2:

- `clang`
- `flang`
- `rust`
- `crystal`
- `mesa`

This package compiled for me without modifications and without errors, but unfortunately I don't yet know how exactly to use it:

- `wasi-libc`

This package fails to compile for me with LLVM 21, but I'm not very familiar with Mingw, so maybe someone else should try to bump and fix it for LLVM 21. If nobody else is available to do so, I could try to bump and fix it myself:

- `llvm-mingw-w64-libcompiler-rt`


<details>
<summary>my log of testing clang, flang, rust, crystal and mesa</summary>

```
~/code $ clang --version
clang version 21.1.2
Target: aarch64-unknown-linux-android24
Thread model: posix
InstalledDir: /data/data/com.termux/files/usr/bin
~/code $ flang --version
flang version 21.1.2
Target: aarch64-unknown-linux-android24
Thread model: posix
InstalledDir: /data/data/com.termux/files/usr/bin
~/code $ crystal --version
Crystal 1.17.1 [19be240] (2025-09-27)

LLVM: 21.1.2
Default target: aarch64-unknown-linux-android
~/code $ rustc --version --verbose
rustc 1.89.0 (29483883e 2025-08-04) (built from a source tarball)
binary: rustc
commit-hash: 29483883eed69d5fb4db01964cdf2af4d86e9cb2
commit-date: 2025-08-04
host: aarch64-linux-android
release: 1.89.0
LLVM version: 21.1.2
~/code $ cat > hello.c << 'EOF'
> #include <stdio.h>
> int main(void) {
>     printf("hello world\n");
> }
> EOF
~/code $ clang hello.c -ohello -fsanitize=address
~/code $ ./hello
hello world
~/code $ cat > hello.cpp << 'EOF'
> #include <iostream>
> int main(void) {
>     std::cout << "hello world" << std::endl;
> }
> EOF
~/code $ clang++ hello.cpp -ohello -fsanitize=address
~/code $ ./hello
hello world
~/code $ cat > hello.f90 << 'EOF'
program hello
  print *, 'Hello, World!'                               
end program hello
EOF              
~/code $ flang hello.f90 -o hello
~/code $ ./hello
 Hello, World!
~/code $ echo 'puts "Hello World!"' > hello.cr
~/code $ crystal build hello.cr
~/code $ ./hello 
Hello World!
~/code $ LIBGL_ALWAYS_SOFTWARE=1 termux-x11 -xstartup xfce4-session &
[1] 6170
~/code $ 
(xfce4-session:6192): libxfce4util-WARNING **: 16:59:43.245: Failed to get a ConsoleKit proxy: Could not connect: No such file or directory
** Message: 16:59:43.294: Failed to fetch _NET_NUMBER_OF_DESKTOPS; assuming 1
** Message: 16:59:43.297: Failed to get _NET_WORKAREA; using full screen dimensions
** Message: 16:59:43.297: Failed to fetch _NET_CURRENT_DESKTOP; assuming 0
** Message: 16:59:43.298: Failed to fetch _NET_CURRENT_DESKTOP; assuming 0
** Message: 16:59:44.073: Failed to fetch _NET_CURRENT_DESKTOP; assuming 0

(xfwm4:6217): xfwm4-WARNING **: 16:59:44.249: Unsupported GL renderer (llvmpipe (LLVM 21.1.2, 128 bits)).
Baloo File Indexing has been disabled
I2025-09-27 21:59:45.429969 instance.cpp:1400] Override Enabled Addons: {}
I2025-09-27 21:59:45.430293 instance.cpp:1401] Override Disabled Addons: {}
I2025-09-27 21:59:45.450357 xcbconnection.cpp:199] Connecting to X11 display, display name::0.0.
I2025-09-27 21:59:45.451006 addonmanager.cpp:204] Loaded addon xcb
I2025-09-27 21:59:45.454043 addonmanager.cpp:204] Loaded addon quickphrase
I2025-09-27 21:59:45.455045 keyboard.cpp:226] Attempting to load keyboard from: ["/data/data/com.termux/files/usr/share/xkeyboard-config-2"] Rule: evdev

** (process:6263): WARNING **: 16:59:45.464: XDG_CURRENT_DESKTOP not set, unable to check against enabled desktop list. Starting anyway...
I2025-09-27 21:59:45.465303 addonmanager.cpp:204] Loaded addon keyboard
I2025-09-27 21:59:45.489357 addonmanager.cpp:204] Loaded addon imselector
I2025-09-27 21:59:45.502537 addonmanager.cpp:204] Loaded addon clipboard
I2025-09-27 21:59:45.514161 addonmanager.cpp:204] Loaded addon dbus
I2025-09-27 21:59:45.516554 addonmanager.cpp:204] Loaded addon unicode
I2025-09-27 21:59:45.537387 addonmanager.cpp:204] Loaded addon ibusfrontend
I2025-09-27 21:59:45.543649 addonmanager.cpp:204] Loaded addon dbusfrontend
I2025-09-27 21:59:45.570435 addonmanager.cpp:204] Loaded addon fcitx4frontend
I2025-09-27 21:59:45.581263 addonmanager.cpp:204] Loaded addon xim
qt.core.qobject.connect: QObject::connect(QNetworkInformation, LanLinkProvider): invalid nullptr parameter
qt.core.qobject.connect: QObject::connect(QNetworkInformation, LanLinkProvider): invalid nullptr parameter
qt.bluetooth: Dummy backend running. Qt Bluetooth module is non-functional.
kdeconnect.core: No local bluetooth adapter found
I2025-09-27 21:59:45.728826 inputmethodmanager.cpp:210] Found 742 input method(s) in addon keyboard
W2025-09-27 21:59:45.730527 xcbkeyboard.cpp:402] Could not load XKB rules
Error sending UDP packet: QAbstractSocket::NetworkError
I2025-09-27 21:59:45.741185 addonmanager.cpp:204] Loaded addon kimpanel
I2025-09-27 21:59:45.742885 addonmanager.cpp:204] Loaded addon virtualkeyboard
I2025-09-27 21:59:45.789659 classicui.cpp:90] Created classicui for x11 display::0.0
I2025-09-27 21:59:45.789746 addonmanager.cpp:204] Loaded addon classicui
I2025-09-27 21:59:45.794224 addonmanager.cpp:204] Loaded addon notifications
I2025-09-27 21:59:45.794443 dbusmodule.cpp:816] Service name change: org.fcitx.Fcitx5  :1.13
I2025-09-27 21:59:45.794634 kimpanel.cpp:116] Kimpanel new owner: 
I2025-09-27 21:59:45.794664 virtualkeyboard.cpp:218] VirtualKeyboard new owner: 

** (xfdesktop:6254): WARNING **: 16:59:46.919: Failed to get system bus: Could not connect: No such file or directory

~/code $ DISPLAY=:0 glmark2
=======================================================
    glmark2 2023.01
=======================================================
    OpenGL Information
    GL_VENDOR:      Mesa
    GL_RENDERER:    zink Vulkan 1.0(Turnip Adreno (TM) 612 (MESA_TURNIP))
    GL_VERSION:     4.0 (Compatibility Profile) Mesa 25.2.3
    Surface Config: buf=32 r=8 g=8 b=8 a=8 depth=24 stencil=0 samples=0
    Surface Size:   800x600 windowed
=======================================================
[build] use-vbo=false: FPS: 156 FrameTime: 6.447 ms
[build] use-vbo=true:
(xapp-sn-watcher:6263): GLib-GObject-CRITICAL **: 17:00:15.784: g_object_set: assertion 'G_IS_OBJECT (object)' failed

(xapp-sn-watcher:6263): GLib-GIO-CRITICAL **: 17:00:15.784: g_dbus_interface_skeleton_flush: assertion 'G_IS_DBUS_INTERFACE_SKELETON (interface_)' failed

(xapp-sn-watcher:6263): GLib-GObject-CRITICAL **: 17:00:15.784: g_object_set: assertion 'G_IS_OBJECT (object)' failed

(xapp-sn-watcher:6263): GLib-GIO-CRITICAL **: 17:00:15.784: g_dbus_interface_skeleton_flush: assertion 'G_IS_DBUS_INTERFACE_SKELETON (interface_)' failed

(xapp-sn-watcher:6263): GLib-GObject-CRITICAL **: 17:00:15.785: invalid (NULL) pointer instance

(xapp-sn-watcher:6263): GLib-GObject-CRITICAL **: 17:00:15.785: signal_emit_valist_unlocked: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed

(xfce4-session:6192): GLib-WARNING **: 17:00:15.843: ../src/glib/gmain.c:6154: waitpid(pid:6263) failed: No child processes (10). See documentation of g_child_watch_source_new() for possible causes.
 FPS: 175 FrameTime: 5.723 ms
[texture] texture-filter=nearest: FPS: 165 FrameTime: 6.082 ms
[texture] texture-filter=linear: FPS: 162 FrameTime: 6.196 ms
[texture] texture-filter=mipmap: FPS: 161 FrameTime: 6.244 ms
[shading] shading=gouraud: FPS: 177 FrameTime: 5.672 ms
[shading] shading=blinn-phong-inf: FPS: 155 FrameTime: 6.484 ms
[shading] shading=phong: FPS: 146 FrameTime: 6.873 ms
[shading] shading=cel: FPS: 149 FrameTime: 6.753 ms
[bump] bump-render=high-poly: FPS: 128 FrameTime: 7.863 ms
[bump] bump-render=normals: FPS: 156 FrameTime: 6.425 ms
[bump] bump-render=height: FPS: 162 FrameTime: 6.190 ms
[effect2d] kernel=0,1,0;1,-4,1;0,1,0;: FPS: 161 FrameTime: 6.248 ms
[effect2d] kernel=1,1,1,1,1;1,1,1,1,1;1,1,1,1,1;: FPS: 148 FrameTime: 6.798 ms
[pulsar] light=false:quads=5:texture=false: FPS: 151 FrameTime: 6.629 ms
[desktop] blur-radius=5:effect=blur:passes=1:separable=true:windows=4: FPS: 115 FrameTime: 8.767 ms
[desktop] effect=shadow:windows=4: FPS: 137 FrameTime: 7.344 ms
[buffer] columns=200:interleave=false:update-dispersion=0.9:update-fraction=0.5:update-method=map: FPS: 118 FrameTime: 8.541 ms
[buffer] columns=200:interleave=false:update-dispersion=0.9:update-fraction=0.5:update-method=subdata: FPS: 118 FrameTime: 8.543 ms
[buffer] columns=200:interleave=true:update-dispersion=0.9:update-fraction=0.5:update-method=map: FPS: 125 FrameTime: 8.044 ms
[ideas] speed=duration: FPS: 119 FrameTime: 8.405 ms
[jellyfish] <default>: FPS: 126 FrameTime: 7.995 ms
[terrain] <default>: FPS: 29 FrameTime: 34.724 ms
[shadow] <default>: FPS: 106 FrameTime: 9.495 ms
[refract] <default>: FPS: 49 FrameTime: 20.469 ms
[conditionals] fragment-steps=0:vertex-steps=0: FPS: 135 FrameTime: 7.410 ms
[conditionals] fragment-steps=5:vertex-steps=0: FPS: 140 FrameTime: 7.178 ms
[conditionals] fragment-steps=0:vertex-steps=5: FPS: 138 FrameTime: 7.286 ms
[function] fragment-complexity=low:fragment-steps=5: FPS: 134 FrameTime: 7.499 ms
[function] fragment-complexity=medium:fragment-steps=5: FPS: 136 FrameTime: 7.406 ms
[loop] fragment-loop=false:fragment-steps=5:vertex-steps=5: FPS: 140 FrameTime: 7.168 ms
[loop] fragment-steps=5:fragment-uniform=false:vertex-steps=5: FPS: 132 FrameTime: 7.592 ms
[loop] fragment-steps=5:fragment-uniform=true:vertex-steps=5: FPS: 135 FrameTime: 7.437 ms
=======================================================
                                  glmark2 Score: 134 
=======================================================
~/code $ 
```

</details>
